src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -86,29 +86,32 @@ public void Update(IContainSagaData saga)
 
         void ValidateUniqueProperties(IContainSagaData saga)
         {
-            var sagaMetaData = sagaModel.FindByEntityName(saga.GetType().FullName);
-
-            if (!sagaMetaData.CorrelationProperties.Any()) return;
-
-            var sagasFromSameType = from s in data
-                                    where
-                                        (s.Value.SagaEntity.GetType() == saga.GetType() && (s.Key != saga.Id))
-                                    select s.Value;
-
-            foreach (var storedSaga in sagasFromSameType)
+            var sagaType = saga.GetType();
+            var sagaMetaData = sagaModel.FindByEntityName(sagaType.FullName);
+            var existingSagas = (from s in data
+                where s.Value.SagaEntity.GetType() == sagaType && (s.Key != saga.Id)
+                select s.Value)
+                .ToList();
+            foreach (var correlationProperty in sagaMetaData.CorrelationProperties)
             {
-                foreach (var correlationProperty in sagaMetaData.CorrelationProperties)
+                var uniqueProperty = sagaType.GetProperty(correlationProperty.Name);
+                if (!uniqueProperty.CanRead)
+                {
+                    continue;
+                }
+                var incomingSagaPropertyValue = uniqueProperty.GetValue(saga, null);
+                if (incomingSagaPropertyValue == null)
+                {
+                    var message = string.Format("Cannot store saga with id '{0}' since the unique property '{1}' has a null value.", saga.Id, uniqueProperty.Name);
+                    throw new InvalidOperationException(message);
+                }
+
+                foreach (var storedSaga in existingSagas)
                 {
-                    var uniqueProperty = saga.GetType().GetProperty(correlationProperty.Name);
-                    if (!uniqueProperty.CanRead)
-                    {
-                        continue;
-                    }
-                    var inComingSagaPropertyValue = uniqueProperty.GetValue(saga, null);
                     var storedSagaPropertyValue = uniqueProperty.GetValue(storedSaga.SagaEntity, null);
-                    if (inComingSagaPropertyValue.Equals(storedSagaPropertyValue))
+                    if (Equals(incomingSagaPropertyValue, storedSagaPropertyValue))
                     {
-                        var message = string.Format("Cannot store a saga. The saga with id '{0}' already has property '{1}' with value '{2}'.", storedSaga.SagaEntity.Id, uniqueProperty, storedSagaPropertyValue);
+                        var message = string.Format("Cannot store a saga. The saga with id '{0}' already has property '{1}' with value '{2}'.", storedSaga.SagaEntity.Id, uniqueProperty.Name, storedSagaPropertyValue);
```

Comment:
Should we be outputting in an exception the value of a property ? Could it be considered a security flow ?